### ISSUE TITLE
Loading VST presets that aren't FXP format

### ISSUE THREAD

**Issue by DBraun** [DBraun/DawDreamer#23]
Some kinds of VSTs don't use the FXP file format. What can be done to overcome this and load them? Sometimes they're human-readable XML files, which can be parsed in Python. Other times they're not, and it would be more challenging.

**Comment by dimentorium** [comment]
Hi DBraun, 
I hope this project is still alive. Just saw this issue and I already wanted to ask if you could support loading of binary chunks to plugins. If thats working, I could help loading VSTPreset and NKS files, see here.
https://github.com/dimentorium/PresetManager/tree/master/PresetManager/Importer

**Comment by DBraun** [comment]
Yes the project is alive :). It seems you know the details of Kontakt presets. That would be a great thing to get working with DawDreamer by any means, either in Python or C++/JUCE. I think [slakh-generation](https://github.com/ethman/slakh-generation/blob/e6454eb57a3683b99cdd16695fe652f83b75bb14/utils.py#L22) did it by replacing the default preset file and *then* loading the instrument. I hope something better is possible.

Other preset files are XML, and it makes sense to parse them like you've done.

**Comment by DBraun** [comment]
Suppose I have a VST with a complicated UI and no ability to save a preset file. Even in this situation, something like Ableton can save the preset, and that's why you can re-open a project and see the user-modified preset. Something like [getStateInformation](https://docs.juce.com/master/tutorial_audio_processor_value_tree_state.html) is probably being used to save the state and pack it into the Ableton project file. However, the content of that state information isn't shown to the user. It would be great to save out these files, but it would require adding a UI as you discussed [here](https://github.com/DBraun/DawDreamer/issues/24#issue-813293762). One could still explore the getStateInformation features before the UI feature is added.

**Comment by dimentorium** [comment]
Yes, there are functions that write out the whole state of the plugin as a binary chunk. This is then used by DAW or other tools to restore the complete setting of the plugin. VSTPreset, FXP and NKS use this binary chunk as well for storing or configuring the plugin. In the VST API this was setchunk and getchunk.
SO in theory, if its possible to add these functions in DAWdreamer, then we could use my functions to read the chunk from the preset file and load it into the plugin. 
Here is also a thread to this https://forum.juce.com/t/loading-a-binary-file-as-a-fxp-into-a-vst-setchunk/4449
With that you could programmatically save and load the whole state of a plugin. I will look tomorrow into my small plugin how this is handled there.

**Comment by dimentorium** [comment]
If I understand everything correctly.
In the pluginProcessor the functions 

processor->getStateInformation (state);      
https://docs.juce.com/master/classAudioProcessor.html#a5d79591b367a7c0516e4ef4d1d6c32b2

processor->setStateInformation (state.getData(), (int) state.getSize()); 
https://docs.juce.com/master/classAudioProcessor.html#a6154837fea67c594a9b35c487894df27

are needed. They are also used in the host example (grapheditorpanel.cpp). These should call the plugin function getstate/setstate thus saving the whole state of the plugin.

**Comment by dimentorium** [comment]
Think I need your help ;-)
This is basically the function for getting the state of the plugin. This works, but I think the way I am getting the data is simply wrong (the magic byte in python is not correct). And I am not deep into c++ to know how to convert the memoryblock into a string for pythong

std::string
PluginProcessor::getState()
{
    if (!myPlugin) {
        std::cout << "Please load the plugin first!" << std::endl;
        return "";
    }
    MemoryBlock state;
    myPlugin->getStateInformation(state);

    //return (state.toString()).toStdString();
    std::string state_info = (Base64::toBase64(state.toBase64Encoding())).toStdString();
    return state_info;
}

**Comment by DBraun** [comment]
That seems like a good start. Next you need to be able to load the state from python.

For convenience it would also be nice to have a method so that you don't have to use python to think about file saving

`synth.save_state("C:/some/path")` 
is better than
```python
test_state = synth.get_state()
with open('C:/some/path', 'w') as f:
    f.write(test_state)
```

I wonder if the way I implemented automation parameters is totally wrong for this paradigm...

**Comment by dimentorium** [comment]
Here is now a better way for the saving of presets
py::bytes
PluginProcessor::getState()
{
    if (!myPlugin) {
        std::cout << "Please load the plugin first!" << std::endl;
        return "";
    }
    MemoryBlock state;
    myPlugin->getStateInformation(state);    
    
    return std::string((const char*)state.getData(), state.getSize());
}

important is the py::bytes as otherwise it results in errors due to non printable characters. Personally would prefer getting the string instead of saving to file, because for other filetypes you will then need to convert first. Or is it possible to make synth.state a member that can be read and written normally but has a save function?

Concerning load, I do not get it running. Do you have a way to debug the dll directly?
py::bytes
PluginProcessor::setState(std::string& data)
{
    if (!myPlugin) {
        std::cout << "Please load the plugin first!" << std::endl;
        return false;
    }
    std::string test = std::string(data);
    MemoryBlock state;
    state.copyFrom((const char*)&test, 0, 4);
    myPlugin->setStateInformation(state.getData(), state.getSize());

    MemoryBlock check_state;
    myPlugin->getStateInformation(check_state);

    return std::string((const char*)state.getData(), state.getSize());
}

I do not get the data correctly into the memory block (no c++ programmer at all). And I did not get the dll to be able to debug everything, so its just try and error.

What do you mean by automation parameters and the paradigm?

**Comment by DBraun** [comment]
pybind11 allows for properties so it's possible to set up both
`synth.state = something` and `state = synth.state`.

It's not a big deal, but I might prefer keeping the getter and setter separate.

I don't have a great strategy for debugging unfortunately.

Why are you hard-coding 4 as the number of bytes?

About the automation parameters, note that every Processor class can implement getStateInformation/setStateInformation https://github.com/DBraun/DawDreamer/blob/758e2a76d9b16cfc38693a1103b6522907d5bce6/Source/ProcessorBase.h#L43-L44

But currently none of the classes actually implement it. For example it's missing from ReverbProcessor https://github.com/DBraun/DawDreamer/blob/master/Source/ReverbProcessor.h

and obviously from PluginProcessor https://github.com/DBraun/DawDreamer/blob/master/Source/PluginProcessor.cpp which is where I assume you're testing these new methods.

So ideally once get/set stateinformation is implemented for PluginProcessor, it could be implemented for all processors. And then we could call get/set on myMainProcessorGraph https://github.com/DBraun/DawDreamer/blob/758e2a76d9b16cfc38693a1103b6522907d5bce6/Source/RenderEngine.h#L51

This might allow you to load an entire graph of processors!
```python
engine = daw.RenderEngine(SAMPLE_RATE, BUFFER_SIZE)
engine.load_state("path/to/state")
```

That feature would make some others important, such as accessing the processors of a loaded graph.

What could be complicated is implementing get/state state information for the custom processors, such as Reverb, because they use a funky automate parameters class https://github.com/DBraun/DawDreamer/blob/master/Source/CustomParameters.h and https://github.com/DBraun/DawDreamer/blob/master/Source/ReverbProcessor.h#L94-L104 for example. Maybe it would require turning these custom automate parameters into an XML structure for saving. Each processor would be similar in doing this.

**Comment by dimentorium** [comment]
So managed to set VS2019 in a way I can debug Python and the C++ code in parallel to see whats happening.

These are the functions for the state, they definately work, I just do not get Tyrell to update after setting the state.

py::bytes
PluginProcessor::getState()
{
    if (!myPlugin) {
        std::cout << "Please load the plugin first!" << std::endl;
        return "";
    }
    MemoryBlock state;
    myPlugin->getStateInformation(state);    
    
    return std::string((const char*)state.getData(), state.getSize());
}

py::bytes
PluginProcessor::setState(std::string& data)
{
    if (!myPlugin) {
        std::cout << "Please load the plugin first!" << std::endl;
        return false;
    }

    MemoryBlock state = MemoryBlock(data.c_str(), data.size());
    myPlugin->setStateInformation((const char*)state.getData(), (int)state.getSize());

    MemoryBlock check_state;
    myPlugin->getStateInformation(check_state);

    return std::string((const char*)check_state.getData(), check_state.getSize());
}

I put these into PluginProcessor.cpp with the according wrapper and binding functions. So yes, it should be possible to save and load a complete graph, this is what I guess most DAWs do as well. What you save in the state doesn`t matter, as this is only used by the plugins, so this is up to you.

The 4 was just for debugging purpose.

Will have to find out why tyrell is not loading the state and check with some other instruments.

**Comment by DBraun** [comment]
Great can you teach me how you’re debugging the DLL with Visual Studio?

**Comment by dimentorium** [comment]
Hope I do not forget too much. See also here https://docs.microsoft.com/de-de/visualstudio/python/debugging-mixed-mode-c-cpp-python-in-visual-studio?view=vs-2019

- Install Python for Visual Studio 2019
- Install Python Debug Symbols https://docs.microsoft.com/de-de/visualstudio/python/debugging-symbols-for-mixed-mode-c-cpp-python?view=vs-2019
- In the DawDreamer Solution create a python Console Application

Python Console Application
- In the references, add the DawDreamer_DynamicLibrary
- Under configuration -> Debugging switch on the Flag "Activate Native Codedebugging"
- Also add the build path of the DawDreamer_DynamicLibrary to search pathes
- Set the Python app as main startup
- Make sure to select the appropriate Env or Interpreter (I used my standard installation)

DawDreamer_DynamicLibrary Project
- Set to Debugging
- Add Prebuild to create a copy from the created dll in same folder with pyd ending. When the DLL is built new, you have to make sure the pyd gets updated, otherwise debugging is not working
- Under Configuration->Debugging set "Python/Native Debugging"

Now you should be able to use your programs directly in VS2019, if you set a breakpoint in Python or the DLL it should stop there. When you start it takes some time because of the symbols that are loaded (I haven`t yet checked how to disable this). You can also jump from a Python call into the dll code directly.

**Comment by virgile-blg** [comment]
Hello !

Is there any follow-up on this implementation for loading presets / instruments from Kontakt ?
Thanks

**Comment by DBraun** [comment]
Not yet. There’s a chance the trick used in slakh-generation could work https://github.com/DBraun/DawDreamer/issues/23#issuecomment-783549253

**Comment by virgile-blg** [comment]
It's working indeed, hacky way though ! Thanks

**Comment by DBraun** [comment]
0.6.2 now has `open_editor()`, `load_state(filepath: str)`, and `save_state(filepath: str)` for Plugin Processor. However, I haven't tested Kontakt. See `tests/test_plugins.py`

**Comment by kyungyunlee** [comment]
@virgile-blg Hi, did the hacky way work on macos?

**Comment by virgile-blg** [comment]
Hi !
Yes, the trick worked well on MacOs. Unfortunately not on Linux as NI VSTs does not support it.
You would just need to re-init the synth each time you would want to switch instrument, by copying it and making it the default instrument of Kontakt.

**Comment by kyungyunlee** [comment]
@virgile-blg Hey, thanks for the quick reply. 
Could you share a snippet of your code related to this ? I think I am not quite understanding how this works..I am getting a seg fault :'(

**Comment by virgile-blg** [comment]
@kyungyunlee Here is a gist with utility classes that I used to synthesize batches of MIDI :

https://gist.github.com/virgile-blg/97d97c9e58bae12add15618f6a0ed030

Hope this helps

**Comment by kyungyunlee** [comment]
@virgile-blg Thanks so much!!! I now get what file is being copied to where 😅 But I am still getting a seg fault error when calling the `make_plugin_processor()` function with `libc++abi: terminating` message... Just in case someone else has experienced the same thing.

**Comment by oliveroliverio** [comment]
> Hi DBraun, 
> 
> I hope this project is still alive. Just saw this issue and I already wanted to ask if you could support loading of binary chunks to plugins. If thats working, I could help loading VSTPreset and NKS files, see here.
> 
> https://github.com/dimentorium/PresetManager/tree/master/PresetManager/Importer

Repo is gone. Anyway this can be reposted?
